lib/auth/version.go

```diff
@@ -20,55 +20,101 @@ package auth
+			// TODO(vapopov): DELETE IN v19.0.0 – the last known version should already be migrated to backend storage.
+			// Fallback to local process storage for backward compatibility with previous versions.
+			lastKnownVersion, err := procStorage.GetTeleportVersion(ctx)
+			if trace.IsNotFound(err) {
+				authInfo, err = authinfo.NewAuthInfo(&authinfov1.AuthInfoSpec{TeleportVersion: currentVersion.String()})
+				if err != nil {
+					return trace.Wrap(err)
+				}
+				if _, err := backendStorage.CreateAuthInfo(ctx, authInfo); err != nil {
+					err = trace.Wrap(err)
+					slog.WarnContext(ctx, "Failed to create AuthInfo resource",
+						"error", err)
+					continue
+				}
+				return nil
+			} else if err != nil {
+				return trace.Wrap(err)
+			}
+
+			authInfo, err = authinfo.NewAuthInfo(&authinfov1.AuthInfoSpec{TeleportVersion: lastKnownVersion.String()})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			if _, err := backendStorage.CreateAuthInfo(ctx, authInfo); err != nil {
+				err = trace.Wrap(err)
+				slog.WarnContext(ctx, "Failed to create AuthInfo resource",
+					"error", err)
+				continue
+			}
+		} else if err != nil {
 			return trace.Wrap(err)
 		}
-		return nil
-	} else if err != nil {
-		return trace.Wrap(err)
-	}
 
-	if currentVersion.Major-lastKnownVersion.Major > 1 {
-		return trace.BadParameter("Unsupported upgrade path detected: from %v to %v. "+
-			"Teleport supports direct upgrades to the next major version only.\n Please upgrade "+
-			"your cluster to version %d.x.x first. See compatibility guarantees for details: "+
-			"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
-			lastKnownVersion, currentVersion.String(), lastKnownVersion.Major+1)
-	}
-	if lastKnownVersion.Major-currentVersion.Major > 1 {
-		return trace.BadParameter("Unsupported downgrade path detected: from %v to %v. "+
-			"Teleport doesn't support major version downgrade.\n Please downgrade "+
-			"your cluster to version %d.x.x first. See compatibility guarantees for details: "+
-			"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
-			lastKnownVersion, currentVersion.String(), lastKnownVersion.Major-1)
-	}
-	if err := storage.WriteTeleportVersion(ctx, currentVersion); err != nil {
-		return trace.Wrap(err)
+		lastKnownVersion, err := semver.NewVersion(authInfo.GetSpec().GetTeleportVersion())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if currentVersion.Major-lastKnownVersion.Major > 1 {
+			return trace.BadParameter("Unsupported upgrade path detected: from %v to %v. "+
+				"Teleport supports direct upgrades to the next major version only.\n Please upgrade "+
+				"your cluster to version %d.x.x first. See compatibility guarantees for details: "+
+				"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
+				lastKnownVersion, currentVersion.String(), lastKnownVersion.Major+1)
+		}
+		if lastKnownVersion.Major-currentVersion.Major > 1 {
```

Comment:
```suggestion
		if lastKnownVersion.Major > currentVersion.Major {
```

Reply:
why? we should restrict only if MajorVersion is higher or lower than 1. 15 -> 16 (allow), 15 -> 17 (deny)

Reply:
This is the downgrade protection, not the upgrade version change limit. We should disallow downgrades across major versions without acknowledging the risk.

Reply:
got you, previously we want to allow major version downgrade

Reply:
Downgrades are not really supported in general, downgrades past a major version even less so. The escape hatch of `--skip-version-check` is always there for those scenarios, right?

Reply:
changed: https://github.com/gravitational/teleport/pull/52837/commits/9b40b98eb237467500cd378ddecc68d2c605a6ba